gpflow/expectations.py

```diff
@@ -364,57 +372,95 @@ def _expectation(p, kern1, feat1, kern2, feat2, nghp=None):
     Compute the expectation:
     expectation[n] = <Ka_{Z1, x_n} Kb_{x_n, Z2}>_p(x_n)
         - Ka_{.,.}, Kb_{.,.} :: RBF kernels
-    Ka and Kb as well as Z1 and Z2 can differ from each other, but this is supported
-    only if the Gaussian p is Diagonal (p.cov NxD) and Ka, Kb have disjoint active_dims
-    in which case the joint expectations simplify into a product of expectations
+    Ka and Kb as well as Z1 and Z2 can differ from each other.
 
-    :return: NxMxM
+    :return: N x dim(Z1) x dim(Z2)
     """
-    if kern1.on_separate_dims(kern2) and isinstance(p, DiagonalGaussian):  # no joint expectations required
+    if kern1.on_separate_dims(kern2) and isinstance(p, DiagonalGaussian):
+        # no joint expectations required
         eKxz1 = expectation(p, (kern1, feat1))
         eKxz2 = expectation(p, (kern2, feat2))
         return eKxz1[:, :, None] * eKxz2[:, None, :]
 
-    if feat1 != feat2 or kern1 != kern2:
-        raise NotImplementedError("The expectation over two kernels has only an "
-                                  "analytical implementation if both kernels are equal.")
+    Ka, Kb = kern1, kern2
 
-    kern = kern1
-    feat = feat1
+    with params_as_tensors_for(Ka, feat1, Kb, feat2):
 
-    with params_as_tensors_for(kern), params_as_tensors_for(feat):
         # use only active dimensions
-        Xcov = kern._slice_cov(tf.matrix_diag(p.cov) if isinstance(p, DiagonalGaussian) else p.cov)
-        Z, Xmu = kern._slice(feat.Z, p.mu)
+        Xcov = Ka._slice_cov(tf.matrix_diag(p.cov) if isinstance(p, DiagonalGaussian) else p.cov)
+
+        Z1, Xmu = Ka._slice(feat1.Z, p.mu)
 
         N = tf.shape(Xmu)[0]
         D = tf.shape(Xmu)[1]
 
-        squared_lengthscales = kern.lengthscales ** 2. if kern.ARD \
-            else tf.zeros((D,), dtype=settings.tf_float) + kern.lengthscales ** 2.
+        def get_squared_length_scales(kern):
+            squared_lengthscales = kern.lengthscales ** 2. if kern.ARD \
+                        else tf.zeros((D,), dtype=settings.tf_float) + kern.lengthscales ** 2.
+            return squared_lengthscales
 
-        sqrt_det_L = tf.reduce_prod(0.5 * squared_lengthscales) ** 0.5
-        C = tf.cholesky(0.5 * tf.matrix_diag(squared_lengthscales) + Xcov)  # NxDxD
+        if Ka == Kb:
```

Comment:
For understanding correctly: does this mean "equal lengthscale parameters"?

Reply:
Yes, equal length scales, we use variances at the end only.